server/store/datastore/permission.go

```diff
@@ -29,19 +31,26 @@ func (s storage) PermFind(user *model.User, repo *model.Repo) (*model.Perm, erro
 }
 
 func (s storage) PermUpsert(perm *model.Perm) error {
-	if len(perm.Repo) == 0 && perm.RepoID == 0 {
-		return fmt.Errorf("could not determine repo for permission: %v", perm)
-	}
-
 	sess := s.engine.NewSession()
 	defer sess.Close()
 	if err := sess.Begin(); err != nil {
 		return err
 	}
 
-	// lookup repo based on name if possible (preserve old behaviour)
-	// TODO: check if needed
-	if len(perm.Repo) != 0 {
+	if err := s.permUpsert(sess, perm); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+func (s storage) permUpsert(sess *xorm.Session, perm *model.Perm) error {
```

Comment:
What was the reason for creating a new function here?

Reply:
to have it within one sql transaction (**`sess *xorm.Session`**) - it's a private function so it's fine to do so :)